Adds robot status publishing to Rapid driver

yamokosk opened the pull request:
Continuation of #144. Fully tested on hardware and a packet capture is included to verify. In capture robot was commanded to move to a few points and at the end of the capture, the E-stop was pressed.

The PR also standardizes whitespace indentation. Spaces, as opposed to tabs, were used more often for indentation so I changed everything to spaces. They are in separate commits to better see the changes.

[robot_status_message.pcapng.zip](https://github.com/ros-industrial/abb/files/1771700/robot_status_message.pcapng.zip)

yamokosk committed:
Adds RobotStatus message publishing to the ABB Rapid driver

yamokosk committed:
Standarizes whitespace indentation to 4 spaces for Rapid code

Four spaces was used more frequently for indentation throughout the Rapid code. Tabs were converted to spaces and in a few places extra spaces were added or removed.

yamokosk committed:
Fixes more whitespace changes.

gavanderhoorn commented:
@Levi-Armstrong: is this something you'd pick up?

@yamokosk: my apologies for our late response, this is definitely something that is of great interest :+1:

gonzalocasas commented:
It looks good to me.

Is there anything I could do to speed up the merge of this PR? I can offer to test it on some of our hardware if that helps. (or if there's another way to contribute to make the process smoother/faster, please let me know).

gavanderhoorn commented:
Getting a second user to verify this on their hw would be good, yes.

If you would be willing to do that, it would be appreciated.

gonzalocasas commented:
Yes, sure, I will test this next week in our hardware and post here! Thx!

gavanderhoorn commented:
@gonzalocasas: have you had a chance to test this?

gonzalocasas commented:
@gavanderhoorn: preparations too a bit longer than expected, but I've scheduled the test for this Tuesday.

gonzalocasas commented:
@yamokosk is this how the signals should be configured?

![image](https://user-images.githubusercontent.com/933277/39710347-deec536a-521c-11e8-8d5c-72841a221a1b.png)

(In general, if a PR requires config changes, I'd suggest to include them as part of the description, so testing/review and transfer to the wiki is unambiguous).

gonzalocasas commented:
Ok, this seems to work.
I am still not sure if the signal configuration that I used is correct, but the `robot_status` looks like properly reflecting it, so lgtm.

gavanderhoorn commented on abb_driver/rapid/ROS_stateServer.mod:
```diff
@@ -37,42 +37,120 @@ LOCAL VAR socketdev client_socket;
+    message.error_code := ERRNO;
+
+    ! Determine in_error
+    if (message.error_code >= 1) AND (message.error_code <= 90) THEN
```
It's been a while for me: are there no errors with `ERRNO > 90`?

gavanderhoorn replied:
Unless I'm misunderstanding something, errors can have nrs `> 90`, so this line may not be sufficient for detecting whether the system is in error or not.

gonzalocasas replied:
I assume that comes out of the following description of `errnum` on RAPID reference:

![image](https://user-images.githubusercontent.com/933277/39763681-1d4708b4-52de-11e8-8dbb-2524bbd4b881.png)

gavanderhoorn replied:
Yes, I had understood that as well. But that bit of doc doesn't seem to limit error nrs to `[1, 90]` specifically.

gavanderhoorn replied:
I'd forgotten about this: a knowledgable source has confirmed that error nrs can definitely be `> 90`, so this implementation is not sufficient to determine controller error status.

Most likely a combination of system outputs would be better.

keerthanamanivannan replied:
@gavanderhoorn do you have any suggestions as to what combination of system outputs might work for this?

gavanderhoorn replied:
I wouldn't know all of them, but I would at least check those that have to do with e-stop status.

Perhaps @jontje can suggest some others?

jontje replied:
Hello,

`errnum = [1, 90]` are reserved for rasing custom errors, while predefined errors can be higher (e.g. `ERR_ROBLIMIT = 1075`).

Anyway, I would use the system output `Execution Error`, and I think it should be enough for determining the `message.in_error` field. This system output is set to `HIGH` when an error has occured that doesn't have any error handler. 

For example:
![system output](https://user-images.githubusercontent.com/16589371/50890566-bf13ac00-13fa-11e9-96f7-39637e5dc368.PNG)

And then check the signal in an IF-statement with `DOutput(signalExecutionError) = HIGH`

keerthanamanivannan replied:
@jontje this was well explained, thank you! 
@gavanderhoorn I made a commit with this change. I updated the documentation too. Please review it.

gavanderhoorn replied:
@keerthanamanivannan: b97cbcbe seems to remove the bit where the code retrieves the current error and initialises the `message.error_code` field with it. Was that intentional?

That field should be set to whatever the current error code is when the system is in an error mode.

keerthanamanivannan replied:
@gavanderhoorn ahh I see. Sorry. I've fixed it.

gavanderhoorn commented:
Thanks for the check @gonzalocasas :+1: 

We'll have to make sure to document the signal configuration, as it's unclear right now, as you remark.

Ideally with both a comment at the top of the file as well as in the installation tutorial.

gavanderhoorn commented:
I actually believe `signalMotorOn` should be mapped to `Motors On State`, not `Motors On`.

gonzalocasas commented:
@gavanderhoorn it could be; but then it's unclear to me what's the correct mapping for `signalMotionPossible`.

gavanderhoorn commented:
Not sure either right now. Not near a robot/robotstudio/manual.

In any case: `drives_powered` (in the ROS msg) should reflect whether the drives are powered *now*. IIRC, `Motors On` is a pulsing signal, it does not stay high when motors are already enabled/on.

gonzalocasas commented:
hi @yamokosk! could you pls give me a hand in describing how you configured the signals in your setup? I can then help with the documentation part so that this PR gets merged ;)

yamokosk commented:
Hi guys, sorry for the delayed response. Sure. Do you want me to document how I created the signals in RobotStudio? I can send screenshots of the relevant screens if that works.

gonzalocasas commented:
@yamokosk yep, screenshots would help

gavanderhoorn commented:
@yamokosk: it would be great if we could include some documentation on how to setup everything needed wrt signals needed for this extension.

Could you provide the screenshot(s) that you offered?

gavanderhoorn commented:
@yamokosk: :bellhop_bell:

gavanderhoorn commented:
And another ping.

keerthanamanivannan commented:
Alright, I work with @yamokosk and I've been working on our robots with these changes. I'm attaching the screenshots that @yamokosk mentioned above. 

![pic1](https://user-images.githubusercontent.com/15019704/47097456-0bce5900-d1f7-11e8-8089-819b94572df9.png)
The above screenshot shows the place where we add these signals.

![pic2](https://user-images.githubusercontent.com/15019704/47097471-1983de80-d1f7-11e8-9c29-1a3dff2795d7.png)
And this one shows what System Outputs we tied the Signals to. 
Let me know if anything else is required!

keerthanamanivannan commented:
@gavanderhoorn @gonzalocasas ping.

gavanderhoorn commented:
> Let me know if anything else is required!

thanks for adding this @keerthanamanivannan.

It would be great if we could add this to either the readme of `abb_driver` and/or the [installation tutorial](http://wiki.ros.org/abb/Tutorials/InstallServer).

keerthanamanivannan commented:
@gavanderhoorn Sure, I can do that! 
How would I edit this link: [installation tutorial](http://wiki.ros.org/abb/Tutorials/InstallServer)?
It says that its an immutable page.

keerthanamanivannan commented:
@gavanderhoorn Sure, I can do that!
How would I edit this link: [installation tutorial](http://wiki.ros.org/abb/Tutorials/InstallServer)?
It says that its an immutable page.

gavanderhoorn commented:
Do you have a wiki account?

keerthanamanivannan commented:
Yes, I do.

gavanderhoorn commented:
And have you requested write access? That's a separate step.

If the answer is also "yes", then you should be able to open [this link](http://wiki.ros.org/action/edit/abb/Tutorials/InstallServer?action=edit&editor=text) and start editing.

keerthanamanivannan commented:
Oh. Well, no. I have never requested write access. How would I do that?

gavanderhoorn commented:
You can request that over in https://github.com/ros-infrastructure/roswiki/issues/258.

gavanderhoorn commented:
@keerthanamanivannan ?

keerthanamanivannan commented:
@gavanderhoorn I edited the abb InstallServer tutorial and clicked the save button and it got published! Thought it might go through some approval process of some sort, but yeah. Review that and change/modify it if necessary. Merge this PR. Let me know if anything else is required.

gavanderhoorn commented:
> I edited the abb InstallServer tutorial and clicked the save button and it got published!

great :) thanks.

> Thought it might go through some approval process of some sort, but yeah.

No, the wiki has no review process :)

> Review that and change/modify it if necessary. Merge this PR. Let me know if anything else is required.

Thanks for updating the docs.

I'll take a look at the PR, review your wiki changes and then merge.

gavanderhoorn commented:
@keerthanamanivannan @yamokosk: I've just tested this PR.

Basic functionality is ok: :+1: for that.

Three observations:

 - minor: robot status msgs are published at the same frequency as joint state msgs
 - robot `mode` is `-1` for me when the controller is in `OP_MAN_PROG` mode
 - `motion_possible` does not seem to take motor state or the state of `T_ROB1` into account: if I switch to `auto` mode, but motors are off, `motion_possible` is `1`. If I switch to `auto`, turn on the motors, but don't (re)start `T_ROB1`, `motion_possible` is also `1`. Without motor power or a running `T_ROB1` the robot will not move, so `motion_possible` should stay `0`

It's been a while since you submitted this PR, so would you still be up for fixing this?

@keerthanamanivannan: :+1: for updating the documentation. The instructions are clear and I encountered no issues following them.

gavanderhoorn reviewed (changes requested):
@keerthanamanivannan @yamokosk: I've just tested this PR.

Basic functionality is ok: :+1: for that.

<del>Three</del> Five observations:

 - minor: robot status msgs are published at the same frequency as joint state msgs (something like 10 Hz would probably be enough)
 - robot `mode` is `-1` (ie: `UNKNOWN`) for me when the controller is in `OP_MAN_PROG` mode (should it be: `CASE OP_MAN_PROG, OP_MAN_TEST:`?)
 - `motion_possible` does not seem to take motor state or the state of `T_ROB1` into account: if I switch to `auto` mode, but motors are off, `motion_possible` is `1`. If I switch to `auto`, turn on the motors, but don't (re)start `T_ROB1`, `motion_possible` is also `1`. Without motor power or a running `T_ROB1` the robot will not move, so `motion_possible` should stay `0`
 - the switch statement mapping RAPID robot mode values to the ROS values does not check for `OP_UNDEF`
 - detection of error states does not seem sufficient (see inline comments on the PR)

It's been a while since you submitted this PR, so would you still be up for fixing this?

@keerthanamanivannan: :+1: for updating the documentation. The instructions are clear and I encountered no issues following them.

keerthanamanivannan committed:
Fixes the OpMode function syntax and add check for OP_UNDEF

keerthanamanivannan committed:
Adds check for Robot-Active and Motors-On signals for Motion Possible to be set to true

keerthanamanivannan commented:
@gavanderhoorn alright, so I made some commits and those should take care of points 2, 3 and 4 of the requested changes. Updated the [documentation](http://wiki.ros.org/abb/Tutorials/InstallServer) too!

I had a question about the first one though: [this line](https://github.com/yamokosk/abb/blob/feature/robot_status_message/abb_driver/rapid/ROS_stateServer.mod#L32) says that the broadcast rate is in fact 10 Hz. and we're waiting for this `update_rate` time to send both `joint_states` and `robot_status` messages [here](https://github.com/yamokosk/abb/blob/feature/robot_status_message/abb_driver/rapid/ROS_stateServer.mod#L44-L46). Correct me if I'm wrong. 

And the fifth one, I'm waiting to hear from @jontje for suggestions.

gavanderhoorn commented:
> @gavanderhoorn alright, so I made some commits and those should take care of points 2, 3 and 4 of the requested changes.

Great. Thanks for that.

re: pt 3: does `signalRobotActive` in 17c1a5de check for the status of `T_ROB1`?

> I had a question about the first one though: [this line](https://github.com/yamokosk/abb/blob/feature/robot_status_message/abb_driver/rapid/ROS_stateServer.mod#L32) says that the broadcast rate is in fact 10 Hz. and we're waiting for this `update_rate` time to send both `joint_states` and `robot_status` messages [here](https://github.com/yamokosk/abb/blob/feature/robot_status_message/abb_driver/rapid/ROS_stateServer.mod#L44-L46). Correct me if I'm wrong.

You're correct. My comment was meant to say "right now status and joint state msgs are published at the same frequency, which is not necessarily what we want". But they're both at 10 Hz right now. It would be good to decouple those publication rates, but that would be something for another PR.

> And the fifth one, I'm waiting to hear from @jontje for suggestions.

Yes. I would have to check documentation and my notes again to answer that one, so if @jontje could comment on that one it would save me some time.

---

Thanks for updating the PR again.

keerthanamanivannan commented:
> re: pt 3: does `signalRobotActive` in [17c1a5d](https://github.com/ros-industrial/abb/commit/17c1a5de357d2526a0739ba049611862b58735f0) check for the status of `T_ROB1`?

I saw [this](http://serioussurvivor.com/wp-content/uploads/2017/03/System-Parameters-3HAC17076-1_revG_en.pdf) doc, pg 253 in the manual itself titled `Mechanical Unit Active` and I believe this System Output would check if T_ROB1 is active or not. Would be great if someone else can confirm this. 

> 
> You're correct. My comment was meant to say "right now status and joint state msgs are published at the same frequency, which is not necessarily what we want". But they're both at 10 Hz right now. It would be good to decouple those publication rates, but that would be something for another PR.

Alright got it, yes, maybe a different PR.
>
> Thanks for updating the PR again.

No problem!

gonzalocasas commented:
Minor question: shouldn't we "namespace" the signals that are linked to a specific unit? i.e. `signalRobotActive` and `signalRobotNotMoving`? Otherwise, setting up for a multi-robot system means more departing from standard setup instructions.

keerthanamanivannan committed:
Checks Execution Error System Output to determine whether robot is in error

keerthanamanivannan commented:
> Minor question: shouldn't we "namespace" the signals that are linked to a specific unit? i.e. `signalRobotActive` and `signalRobotNotMoving`? Otherwise, setting up for a multi-robot system means more departing from standard setup instructions.

Good point, and if I had to do this, do you know how I would go about implementing this? 
@gavanderhoorn @jontje any ideas?

@jontje I see a "Assigned to Device" option when I'm creating a Signal in the Robot Controller. But it's showing up empty. Would that be of any help to this task at all?
![capture 3](https://user-images.githubusercontent.com/15019704/50916394-ac03da80-1400-11e9-9432-268f6f428222.PNG)

gavanderhoorn commented:
@keerthanamanivannan: could you please *rebase* your PR on-top of current `kinetic-devel`? Travis refuses to run CI for this PR. I'm not sure whether that is because of the conflicts, but they'll need to be resolved in any case.

keerthanamanivannan committed:
Adds code to set the error_code which was removed by mistake

keerthanamanivannan committed:
Changes the error_code only if the system is in_error

keerthanamanivannan committed:
Adds a minor formatting change

keerthanamanivannan commented:
@gavanderhoorn Please review. Thanks!

jontje commented:
> Good point, and if I had to do this, do you know how I would go about implementing this?
> @gavanderhoorn @jontje any ideas?

When I have multiple mechanical units in a system then I usually append the unit names as a prefix.
For example using "ROB_X" like this: ROB_1_signal_name, ROB_2_signal_name, etc...

I would also choose to create a RobotWare Add-In that inspects the robot system during installation, and automatically loads the necessary configurations and RAPID modules. However, this would probably require quite a bit of effort if you are not familiar with creating such Add-Ins.

> @jontje I see a "Assigned to Device" option when I'm creating a Signal in the Robot Controller. But it's showing up empty. Would that be of any help to this task at all?

The "Assigned to Device" field is only required when connecting the signal to a physical IO-device. When the field is left empty, then the signal will only be virtual. So it will not be of any help here.

gavanderhoorn commented:
> > Good point, and if I had to do this, do you know how I would go about implementing this?
> > @gavanderhoorn @jontje any ideas?
> 
> When I have multiple mechanical units in a system then I usually append the unit names as a prefix.
> For example using "ROB_X" like this: ROB_1_signal_name, ROB_2_signal_name, etc...

hm. That would need changes to the code when deploying on systems with multiple mechanical units.

@jontje: there is no way to "scope" these signal declarations that would make it less intrusive? Something similar to assigning to a mech unit, as you do with motion tasks?

keerthanamanivannan commented:
> hm. That would need changes to the code when deploying on systems with multiple mechanicle units.

Right. I actually work on a multi-robot system. I set each robot separately as a single mechanical unit in ABB Settings. Like, all the robots are named `ROB_1` in the ABB settings, but I launch all the nodes for each robot in its own namespace. Like @gavanderhoorn mentioned in #106.

So I essentially get the info on all the robots like `robot1/robot_status` and `robot2/robot_status`, and so on.

jontje commented:
> @jontje: there is no way to "scope" these signal declarations that would make it less intrusive? Something similar to assigning to a mech unit, as you do with motion tasks?

Hm, no direct way that I can think of at the moment. Some of the signals are related to the robot controller (e.g. `signalRobotEStop`) and others are related to mechanical units (e.g. `signalRobotNotMoving`). Maybe an idea could be to use cross connections to apply some logic to the signals related to multiple mechanical units.

> Right. I actually work on a multi-robot system. I set each robot separately as a single mechanical unit in ABB Settings. Like, all the robots are named `ROB_1` in the ABB settings, but I launch all the nodes for each robot in its own namespace. Like @gavanderhoorn mentioned in #106.
> 
> So I essentially get the info on all the robots like `robot1/robot_status` and `robot2/robot_status`, and so on.

If I understand correctly, then you work with a multi robot controller system (with one robot each), instead of one robot controller with multiple robots? If so, to make things as generalized as possible then it might be reasonable to set up some namespaces like this:

- `robot_controller1/robot1/robot_status`
- `robot_controller1/robot2/robot_status`
- `robot_controller1/robot3/robot_status`
- `robot_controller2/robot1/robot_status`
- `robot_controller2/robot2/robot_status`
- Etc...

keerthanamanivannan commented:
> If I understand correctly, then you work with a multi robot controller system (with one robot each), instead of one robot controller with multiple robots? If so, to make things as generalized as possible then it might be reasonable to set up some namespaces like this:
> 
> * `robot_controller1/robot1/robot_status`
> * `robot_controller1/robot2/robot_status`
> * `robot_controller1/robot3/robot_status`
> * `robot_controller2/robot1/robot_status`
> * `robot_controller2/robot2/robot_status`
> * Etc...

Yes you are correct, I do not have the setup of one robot controller with multiple robots. 
I have four robots and four controllers, with one to one mapping. So my namespacing setup works for me. But I hear you, yeah.

keerthanamanivannan commented:
@gavanderhoorn how are we moving forward with this?

gavanderhoorn commented:
@keerthanamanivannan: all your commits appear to not be attributed to any email address Github knows about. Is that ok for you? Otherwise I'd recommend you either register the email address with your account, or update the commits.

keerthanamanivannan commented:
@gavanderhoorn thank you for bringing that to my attention. I added the email address to my account. :+1:

gavanderhoorn committed:
state_server: if not in error, set errcode to 0.

Better aligned with other simple_message server implementations.

gavanderhoorn committed:
state_server: motion_possible depends on state of ROS motion task.

So take that state into account. Only when motion task (not motion server!) is running can motions be executed.

gavanderhoorn committed:
state_server: sort signal defs.

gavanderhoorn commented:
Ok.

So I would like to get this merged in.

Even without explicit "support" for multi-mech-unit setups (or: a nice way to scale/scope the signals in those cases).

I've just added ee970d4 which fixes the `in_motion` field (it didn't take execution status of the motion execution task into account, leading to `in_motion` being `1` while in reality trajectories could not be executed).

I'll wait for Travis to give us a green checkmark and then I'll merge.

gavanderhoorn commented:
I've also updated the [install server](http://wiki.ros.org/action/info/abb/Tutorials/InstallServer?action=diff&rev2=25&rev1=21) tutorial on the wiki to mention the new signal.

gonzalocasas commented:
> I've also updated the [install server](http://wiki.ros.org/action/info/abb/Tutorials/InstallServer?action=diff&rev2=25&rev1=21) tutorial on the wiki to mention the new signal.

👍 

I actually would prefer to move all those instructions to github, so that they evolve together with the codebase (hence https://github.com/ros-industrial/abb/pull/161), but wiki seems to be the place for it right now.

gavanderhoorn commented:
> I actually would prefer to move all those instructions to github, so that they evolve together with the codebase (hence #161), but wiki seems to be the place for it right now.

Yes, that's the still ongoing debate about where ROS pkg documentation should live.

The wiki is still the go-to place for many users. If something is not on the wiki, it doesn't exist for many.

gavanderhoorn commented:
I'm going to close this in favour of #168.

As I wrote in the PR there: all the hard work has been done by @yamokosk and @keerthanamanivannan in this PR. I've merely cleaned up and reordered some things.

Thanks @yamokosk and @keerthanamanivannan :+1:.

Apologies for taking so long and all the iterations.
